File: ch23p815attributedStringDrawing2/ch23p773attributedStringDrawing/StringDrawer.m

#import "StringDrawer.h"

@implementation StringDrawer

- (void) setAttributedText:(NSAttributedString *)attributedText {
    self->_attributedText = [attributedText copy];
    [self setNeedsDisplay];
}

- (void)drawRect:(CGRect)rect
{

#define which 2
    
#if which == 1
    
    CGRect r = CGRectOffset(rect, 0, 2); // shoved down a little from top
    [self.attributedText drawWithRect:r options:NSStringDrawingTruncatesLastVisibleLine | NSStringDrawingUsesLineFragmentOrigin context:nil];

#elif which == 2
    
    NSLayoutManager* lm = [NSLayoutManager new];
    NSTextStorage* ts = [[NSTextStorage alloc] initWithAttributedString:self.attributedText];
    [ts addLayoutManager:lm];
    NSTextContainer* tc =
    [[NSTextContainer alloc]
     initWithSize:rect.size];
    [lm addTextContainer:tc];
    tc.lineBreakMode = NSLineBreakByTruncatingTail;
    tc.lineFragmentPadding = 0;
    NSRange r = [lm glyphRangeForTextContainer:tc];
    [lm drawBackgroundForGlyphRange:r atPoint:CGPointMake(0,2)];
    [lm drawGlyphsForGlyphRange:r atPoint:CGPointMake(0,2)];
    
#endif
    
}

@end

File: iOS11bookExamples/bk2ch10p510attributedStringDrawing2/ch23p773attributedStringDrawing/StringDrawer.swift


import UIKit

class StringDrawer : UIView {
    @NSCopying var attributedText : NSAttributedString! {
        didSet {
            self.setNeedsDisplay()
        }
    }
    
    let which = 2
    
    override func draw(_ rect: CGRect) {
        
        switch which {
        case 1:
            let r = rect.offsetBy(dx: 0, dy: 2)
            let options : NSStringDrawingOptions = [.truncatesLastVisibleLine, .usesLineFragmentOrigin]
            self.attributedText.draw(with:r, options: options, context: nil)
        case 2:
            let lm = NSLayoutManager()
            let ts = NSTextStorage(attributedString:self.attributedText)
            ts.addLayoutManager(lm)
            let tc = NSTextContainer(size:rect.size)
            lm.addTextContainer(tc)
            tc.lineBreakMode = .byTruncatingTail //
            tc.lineFragmentPadding = 0
            let r = lm.glyphRange(for:tc)
            lm.drawBackground(forGlyphRange:r, at:CGPoint(0,2))
            lm.drawGlyphs(forGlyphRange: r, at:CGPoint(0,2))
        default:break
        }
        
    }
}
